File: snowplow-tracker/src/androidTest/java/com/snowplowanalytics/snowplow/tracker/TrackerTest.java
/*
 * Copyright (c) 2015-2020 Snowplow Analytics Ltd. All rights reserved.
 *
 * This program is licensed to you under the Apache License Version 2.0,
 * and you may not use this file except in compliance with the Apache License Version 2.0.
 * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the Apache License Version 2.0 is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
 */

package com.snowplowanalytics.snowplow.tracker;

import android.annotation.SuppressLint;
import android.test.AndroidTestCase;
import android.util.Log;

import com.snowplowanalytics.snowplow.tracker.constants.Parameters;
import com.snowplowanalytics.snowplow.tracker.emitter.BufferOption;
import com.snowplowanalytics.snowplow.tracker.events.ScreenView;
import com.snowplowanalytics.snowplow.tracker.events.Timing;
import com.snowplowanalytics.snowplow.tracker.tracker.ExceptionHandler;
import com.snowplowanalytics.snowplow.tracker.tracker.ScreenState;
import com.snowplowanalytics.snowplow.tracker.utils.LogLevel;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.IOException;
import java.util.Map;

import java.util.UUID;
import java.util.concurrent.TimeUnit;

import okhttp3.mockwebserver.MockResponse;
import okhttp3.mockwebserver.MockWebServer;
import okhttp3.mockwebserver.RecordedRequest;

public class TrackerTest extends AndroidTestCase {

    @Override
    protected void setUp() throws Exception {
        super.setUp();
        try {
            Tracker tracker = Tracker.instance();
            Emitter emitter = tracker.getEmitter();
            emitter.shutdown(30);
            tracker.close();
            emitter.waitForEventStore();
            boolean isClean = emitter.getEventStore().removeAllEvents();
            Log.i("TrackerTest", "Tracker closed - EventStore cleaned: " + isClean);
            Log.i("TrackerTest", "Events in the store: " + emitter.getEventStore().getSize());
        } catch(IllegalStateException e) {
            Log.i("TrackerTest", "Tracker already closed.");
        }
    }

    // Helper Methods
    private Tracker getTracker() {
        return getTracker(false);
    }

    private Tracker getTracker(boolean installTracking) {
        Emitter emitter = new Emitter
                .EmitterBuilder("testUrl", getContext())
                .tick(0)
                .emptyLimit(0)
                .build();
        emitter.waitForEventStore();

        Subject subject = new Subject
                .SubjectBuilder()
                .context(getContext())
                .build();

        return new Tracker.TrackerBuilder(emitter, "myNamespace", "myAppId", getContext())
            .subject(subject)
            .platform(DevicePlatforms.InternetOfThings)
            .base64(false)
            .level(LogLevel.VERBOSE)
            .threadCount(1)
            .sessionContext(false)
            .mobileContext(false)
            .geoLocationContext(false)
            .foregroundTimeout(5)
            .backgroundTimeout(5)
            .sessionCheckInterval(15)
            .timeUnit(TimeUnit.SECONDS)
            .applicationCrash(false)
            .lifecycleEvents(true)
            .installTracking(installTracking)
            .applicationContext(true)
            .build();
    }

    // Tests

    public void testTrackerNotInit() {
        boolean exception = false;
        try {
            Tracker.instance();
        } catch (Exception e) {
            assertEquals("FATAL: Tracker must be initialized first!", e.getMessage());
            exception = true;
        }
        assertTrue(exception);
    }

    public void testSetValues() {
        Tracker tracker = getTracker(true);
        assertEquals("myNamespace", tracker.getNamespace());
        assertEquals("myAppId", tracker.getAppId());
        assertEquals(DevicePlatforms.InternetOfThings, tracker.getPlatform());
        assertEquals(false, tracker.getBase64Encoded());
        assertNotNull(tracker.getEmitter());
        assertNotNull(tracker.getSubject());
        assertEquals(LogLevel.VERBOSE, tracker.getLogLevel());
        assertEquals(2, tracker.getThreadCount());
        assertEquals(false, tracker.getApplicationCrash());
        assertEquals(true, tracker.getLifecycleEvents());
        assertEquals(true, tracker.getInstallTracking());
        assertEquals(true, tracker.getApplicationContext());
    }

    public void testEmitterUpdate() {
        Tracker tracker = getTracker();
        assertNotNull(tracker.getEmitter());

        tracker.setEmitter(new Emitter.EmitterBuilder("test", getContext()).build());
        assertNotNull(tracker.getEmitter());
    }

    public void testSubjectUpdate() {
        Tracker tracker = getTracker();
        assertNotNull(tracker.getSubject());

        tracker.setSubject(null);
        assertNull(tracker.getSubject());
    }

    public void testPlatformUpdate() {
        Tracker tracker = getTracker();
        assertEquals(DevicePlatforms.InternetOfThings, tracker.getPlatform());

        tracker.setPlatform(DevicePlatforms.Mobile);
        assertEquals(DevicePlatforms.Mobile, tracker.getPlatform());
    }

    public void testDataCollectionSwitch() {
        Tracker tracker = getTracker();
        assertTrue(tracker.getDataCollection());

        tracker.pauseEventTracking();
        assertFalse(tracker.getDataCollection());
        tracker.pauseEventTracking();
        assertFalse(tracker.getDataCollection());

        tracker.resumeEventTracking();
        assertTrue(tracker.getDataCollection());
        tracker.resumeEventTracking();
        assertTrue(tracker.getDataCollection());
    }

    public void testTrackEventMultipleTimes() {
        Timing event = Timing.builder()
                .category("category")
                .variable("variable")
                .timing(100)
                .build();
        UUID id1 = new TrackerEvent(event).eventId;
        UUID id2 = new TrackerEvent(event).eventId;
        assertFalse(id1.equals(id2));
    }

    public void testTrackWithNoContext() throws Exception {
        Executor.setThreadCount(30);
        Executor.shutdown();

        MockWebServer mockWebServer = getMockServer(1);

        Emitter emitter = null;
        try {
            emitter = new Emitter.EmitterBuilder(getMockServerURI(mockWebServer), getContext())
                    .option(BufferOption.Single)
                    .build();
            emitter.waitForEventStore();
        } catch (Exception e) {
            e.printStackTrace();
            fail("Exception on Emitter creation");
        }

        Tracker tracker = new Tracker.TrackerBuilder(emitter, "myNamespace", "testTrackWithNoContext", getContext())
                .base64(false)
                .level(LogLevel.VERBOSE)
                .sessionContext(false)
                .mobileContext(false)
                .screenContext(false)
                .geoLocationContext(false)
                .build();

        Log.i("testTrackWithNoContext", "Send ScreenView event");
        tracker.track(ScreenView.builder().build());
        Log.i("testTrackWithNoContext", "allEvents: \n" + tracker.getEmitter().getEventStore().getAllEvents());
        RecordedRequest req = mockWebServer.takeRequest(60, TimeUnit.SECONDS);
        assertNotNull(req);
        int reqCount = mockWebServer.getRequestCount();
        assertEquals(1, reqCount);

        JSONObject payload = new JSONObject(req.getBody().readUtf8());
        assertEquals(2, payload.length());
        assertEquals(
                "iglu:com.snowplowanalytics.snowplow/payload_data/jsonschema/1-0-4",
                payload.getString("schema")
        );

        JSONArray data = payload.getJSONArray("data");
        assertEquals(1, data.length());
        JSONObject event = data.getJSONObject(0);

        boolean found = true;
        try {
            String co = event.getString(Parameters.CONTEXT);
            Log.e("testTrackWithNoContext", "Unexpected event: " + event.toString());
            fail(co);
        } catch (Exception e) {
            found = false;
        }
        assertFalse(found);

        mockWebServer.shutdown();
    }

    public void testTrackWithoutDataCollection() throws Exception {
        MockWebServer mockWebServer = getMockServer(1);

        Emitter emitter = new Emitter.EmitterBuilder(getMockServerURI(mockWebServer), getContext())
                .option(BufferOption.Single)
                .build();
        emitter.waitForEventStore();

        Tracker tracker = new Tracker.TrackerBuilder(emitter, "myNamespace", "myAppId", getContext())
            .base64(false)
            .level(LogLevel.VERBOSE)
            .sessionContext(false)
            .mobileContext(false)
            .geoLocationContext(false)
            .build();

        tracker.pauseEventTracking();
        tracker.track(ScreenView.builder().build());
        RecordedRequest req = mockWebServer.takeRequest(2, TimeUnit.SECONDS);

        assertEquals(0, tracker.getEmitter().getEventStore().getSize());
        assertNull(req);

        mockWebServer.shutdown();
    }

    public void testTrackWithSession() throws Exception {
        MockWebServer mockWebServer = getMockServer(1);

        Emitter emitter = new Emitter.EmitterBuilder(getMockServerURI(mockWebServer), getContext())
                .option(BufferOption.Single)
                .build();
        emitter.waitForEventStore();

        Tracker tracker = new Tracker.TrackerBuilder(emitter, "myNamespace", "myAppId", getContext())
            .base64(false)
            .level(LogLevel.VERBOSE)
            .sessionContext(true)
            .mobileContext(false)
            .geoLocationContext(false)
            .foregroundTimeout(5)
            .backgroundTimeout(5)
            .sessionCheckInterval(1)
            .timeUnit(TimeUnit.SECONDS)
            .build();

        assertNotNull(tracker.getSession());
        tracker.resumeSessionChecking();
        Thread.sleep(2000);
        tracker.pauseSessionChecking();

        mockWebServer.shutdown();
    }

    public void testTrackScreenView() throws Exception {
        Emitter emitter = new Emitter.EmitterBuilder("fake-uri", getContext())
                .option(BufferOption.Single)
                .build();
        emitter.waitForEventStore();

        Tracker tracker = new Tracker.TrackerBuilder(emitter, "myNamespace", "myAppId", getContext())
                .base64(false)
                .level(LogLevel.VERBOSE)
                .sessionContext(false)
                .mobileContext(false)
                .geoLocationContext(false)
                .screenContext(true)
                .foregroundTimeout(5)
                .backgroundTimeout(5)
                .sessionCheckInterval(1)
                .timeUnit(TimeUnit.SECONDS)
                .build();

        ScreenState screenState = tracker.getScreenState();
        assertNotNull(screenState);

        Map<String, Object> screenStateMapWrapper = screenState.getCurrentScreen(true).getMap();
        Map<String, Object> screenStateMap = (Map<String, Object>)screenStateMapWrapper.get(Parameters.DATA);
        assertEquals("Unknown", screenStateMap.get(Parameters.SCREEN_NAME));

        // Send screenView
        ScreenView screenView = ScreenView.builder().name("screen1").build();
        String screenId = (String)screenView.getData().getMap().get("id");
        tracker.track(screenView);

        screenStateMapWrapper = screenState.getCurrentScreen(true).getMap();
        screenStateMap = (Map<String, Object>)screenStateMapWrapper.get(Parameters.DATA);
        assertEquals("screen1", screenStateMap.get(Parameters.SCREEN_NAME));
        assertEquals(screenId, screenStateMap.get(Parameters.SCREEN_ID));

        // Send another screenView
        screenView = ScreenView.builder().name("screen2").build();
        String screenId1 = (String)screenView.getData().getMap().get("id");
        tracker.track(screenView);

        Map<String, Object> payload = (Map<String, Object>)screenView.getPayload().getMap().get("data");
        assertEquals("screen2", payload.get(Parameters.SCREEN_NAME));
        assertEquals(screenId1, payload.get(Parameters.SCREEN_ID));
        assertEquals("screen1", payload.get(Parameters.SV_PREVIOUS_NAME));
        assertEquals(screenId, payload.get(Parameters.SV_PREVIOUS_ID));
    }

    public void testTrackUncaughtException() throws InterruptedException {
        Thread.setDefaultUncaughtExceptionHandler(
                new TestExceptionHandler("Illegal State Exception has been thrown!")
        );

        assertEquals(
                TestExceptionHandler.class,
                Thread.getDefaultUncaughtExceptionHandler().getClass()
        );

        Emitter emitter = new Emitter.EmitterBuilder("com.acme", getContext()).build();
        emitter.waitForEventStore();

        Tracker tracker = new Tracker.TrackerBuilder(emitter, "myNamespace", "myAppId", getContext())
                .base64(false)
                .level(LogLevel.VERBOSE)
                .applicationCrash(true)
                .build();

        assertTrue(tracker.getApplicationCrash());
        assertEquals(
                ExceptionHandler.class,
                Thread.getDefaultUncaughtExceptionHandler().getClass()
        );
    }

    public void testExceptionHandler() {
        TestExceptionHandler handler = new TestExceptionHandler("Illegal State Exception has been thrown!");
        Thread.setDefaultUncaughtExceptionHandler(handler);

        assertEquals(
                TestExceptionHandler.class,
                Thread.getDefaultUncaughtExceptionHandler().getClass()
        );

        Emitter emitter = new Emitter.EmitterBuilder("com.acme", getContext()).build();
        new Tracker.TrackerBuilder(emitter, "myNamespace", "myAppId", getContext())
                .base64(false)
                .level(LogLevel.VERBOSE)
                .applicationCrash(false)
                .build();
        emitter.waitForEventStore();

        ExceptionHandler handler1 = new ExceptionHandler();
        Thread.setDefaultUncaughtExceptionHandler(handler1);

        assertEquals(
                ExceptionHandler.class,
                Thread.getDefaultUncaughtExceptionHandler().getClass()
        );

        handler1.uncaughtException(Thread.currentThread(), new Throwable("Illegal State Exception has been thrown!"));
    }

    public static class TestExceptionHandler implements Thread.UncaughtExceptionHandler {

        private final String expectedMessage;

        public TestExceptionHandler(String message) {
            this.expectedMessage = message;
        }

        public void uncaughtException(Thread t, Throwable e) {
            assertEquals(this.expectedMessage, e.getMessage());
        }
    }

    // Mock Server

    public MockWebServer getMockServer(int count) throws IOException {
        MockWebServer mockServer = new MockWebServer();
        mockServer.start();
        MockResponse mockResponse = new MockResponse().setResponseCode(200);
        for (int i = 0; i < count; i++) {
            mockServer.enqueue(mockResponse);
        }
        return mockServer;
    }

    @SuppressLint("DefaultLocale")
    public String getMockServerURI(MockWebServer mockServer) {
        if (mockServer != null) {
            return String.format("%s:%d", mockServer.getHostName(), mockServer.getPort());
        }
        return null;
    }
}

File: snowplow-tracker/src/androidTest/java/com/snowplowanalytics/snowplow/internal/tracker/TrackerTest.kt
/*
 * Copyright (c) 2015-2022 Snowplow Analytics Ltd. All rights reserved.
 *
 * This program is licensed to you under the Apache License Version 2.0,
 * and you may not use this file except in compliance with the Apache License Version 2.0.
 * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the Apache License Version 2.0 is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
 */
package com.snowplowanalytics.snowplow.internal.tracker

import android.annotation.SuppressLint
import android.content.Context
import android.util.Log
import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.platform.app.InstrumentationRegistry
import com.snowplowanalytics.core.constants.Parameters
import com.snowplowanalytics.core.emitter.Emitter
import com.snowplowanalytics.core.emitter.Executor.shutdown
import com.snowplowanalytics.core.emitter.Executor.threadCount
import com.snowplowanalytics.core.tracker.ExceptionHandler
import com.snowplowanalytics.core.tracker.Subject
import com.snowplowanalytics.core.tracker.Tracker
import com.snowplowanalytics.core.tracker.TrackerEvent
import com.snowplowanalytics.snowplow.TestUtils
import com.snowplowanalytics.snowplow.emitter.BufferOption
import com.snowplowanalytics.snowplow.event.ScreenView
import com.snowplowanalytics.snowplow.event.SelfDescribing
import com.snowplowanalytics.snowplow.event.Structured
import com.snowplowanalytics.snowplow.event.Timing
import com.snowplowanalytics.snowplow.payload.SelfDescribingJson
import com.snowplowanalytics.snowplow.tracker.DevicePlatform
import com.snowplowanalytics.snowplow.tracker.LogLevel
import okhttp3.mockwebserver.MockResponse
import okhttp3.mockwebserver.MockWebServer
import org.json.JSONException
import org.json.JSONObject
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import java.io.IOException
import java.util.concurrent.TimeUnit
import java.util.function.Consumer

@RunWith(AndroidJUnit4::class)
class TrackerTest {
    @Before
    @Synchronized
    @Throws(Exception::class)
    fun setUp() {
        try {
            if (Companion.tracker == null) return
            val emitter = Companion.tracker!!.emitter
            val eventStore = emitter.eventStore
            if (eventStore != null) {
                val isClean = eventStore.removeAllEvents()
                Log.i("TrackerTest", "EventStore cleaned: $isClean")
                Log.i("TrackerTest", "Events in the store: " + eventStore.size())
            } else {
                Log.i("TrackerTest", "EventStore null")
            }
            emitter.shutdown(30)
            Companion.tracker!!.close()
            Log.i("TrackerTest", "Tracker closed")
        } catch (e: IllegalStateException) {
            Log.i("TrackerTest", "Tracker already closed.")
        }
    }

    // Helper Methods
    private val tracker: Tracker?
        get() = getTracker(false)

    @Synchronized
    private fun getTracker(installTracking: Boolean): Tracker? {
        val namespace = "myNamespace"
        TestUtils.createSessionSharedPreferences(context, namespace)
        val builder = Consumer { emitter: Emitter ->
            emitter.emitterTick = 0
            emitter.emptyLimit = 0
        }
        val emitter = Emitter(
            context, "testUrl", builder
        )
        val subject = Subject(
            context, null
        )
        val trackerBuilder = Consumer { tracker: Tracker ->
            tracker.subject = subject
            tracker.platform = DevicePlatform.InternetOfThings
            tracker.base64Encoded = false
            tracker.logLevel = LogLevel.VERBOSE
            tracker.threadCount = 1
            tracker.sessionContext = false
            tracker.platformContextEnabled = false
            tracker.geoLocationContext = false
            tracker.foregroundTimeout = 5
            tracker.backgroundTimeout = 5
            tracker.timeUnit = TimeUnit.SECONDS
            tracker.exceptionAutotracking = false
            tracker.lifecycleAutotracking = true
            tracker.installAutotracking = installTracking
            tracker.applicationContext = true
        }
        Companion.tracker = Tracker(emitter, "myNamespace", "myAppId", context, trackerBuilder)
        return Companion.tracker
    }

    private val context: Context
        get() = InstrumentationRegistry.getInstrumentation().targetContext

    // Tests
    @Test
    fun testSetValues() {
        val tracker = getTracker(true)
        Assert.assertEquals("myNamespace", tracker!!.namespace)
        Assert.assertEquals("myAppId", tracker.appId)
        Assert.assertEquals(DevicePlatform.InternetOfThings, tracker.platform)
        Assert.assertFalse(tracker.base64Encoded)
        Assert.assertNotNull(tracker.emitter)
        Assert.assertNotNull(tracker.subject)
        Assert.assertEquals(LogLevel.VERBOSE, tracker.logLevel)
        Assert.assertEquals(2, tracker.threadCount.toLong())
        Assert.assertFalse(tracker.exceptionAutotracking)
        Assert.assertTrue(tracker.lifecycleAutotracking)
        Assert.assertTrue(tracker.installAutotracking)
        Assert.assertTrue(tracker.applicationContext)
    }

    @Test
    fun testEmitterUpdate() {
        val tracker = tracker
        Assert.assertNotNull(tracker!!.emitter)
        tracker.emitter = Emitter(context, "test", null)
        Assert.assertNotNull(tracker.emitter)
    }

    @Test
    fun testSubjectUpdate() {
        val tracker = tracker
        Assert.assertNotNull(tracker!!.subject)
        tracker.subject = null
        Assert.assertNull(tracker.subject)
    }

    @Test
    fun testPlatformUpdate() {
        val tracker = tracker
        Assert.assertEquals(DevicePlatform.InternetOfThings, tracker!!.platform)
        tracker.platform = DevicePlatform.Mobile
        Assert.assertEquals(DevicePlatform.Mobile, tracker.platform)
    }

    @Test
    fun testDataCollectionSwitch() {
        val tracker = tracker
        Assert.assertTrue(tracker!!.dataCollection)
        tracker.pauseEventTracking()
        Assert.assertFalse(tracker.dataCollection)
        tracker.pauseEventTracking()
        Assert.assertFalse(tracker.dataCollection)
        tracker.resumeEventTracking()
        Assert.assertTrue(tracker.dataCollection)
        tracker.resumeEventTracking()
        Assert.assertTrue(tracker.dataCollection)
    }

    @Test
    fun testTrackEventMultipleTimes() {
        val event = Timing("category", "variable", 100)
        val id1 = TrackerEvent(event).eventId
        val id2 = TrackerEvent(event).eventId
        Assert.assertNotEquals(id1, id2)
    }

    @Test
    @Throws(JSONException::class, IOException::class, InterruptedException::class)
    fun testTrackSelfDescribingEvent() {
//        Executor.setThreadCount(30);
//        Executor.shutdown();
        val namespace = "myNamespace"
        TestUtils.createSessionSharedPreferences(context, namespace)
        val mockWebServer = getMockServer(1)
        var emitter: Emitter? = null
        val builder =
            Consumer { emitterArg: Emitter -> emitterArg.bufferOption = BufferOption.Single }
        try {
            emitter = Emitter(context, getMockServerURI(mockWebServer)!!, builder)
        } catch (e: Exception) {
            e.printStackTrace()
            Assert.fail("Exception on Emitter creation")
        }
        val trackerBuilder = Consumer { tracker: Tracker ->
            tracker.base64Encoded = false
            tracker.logLevel = LogLevel.VERBOSE
            tracker.sessionContext = false
            tracker.platformContextEnabled = false
            tracker.screenContext = false
            tracker.geoLocationContext = false
            tracker.installAutotracking = false
            tracker.screenViewAutotracking = false
        }
        Companion.tracker =
            Tracker(emitter!!, namespace, "testTrackWithNoContext", context, trackerBuilder)
        val eventStore = emitter.eventStore
        if (eventStore != null) {
            val isClean = eventStore.removeAllEvents()
            Log.i("testTrackSelfDescribingEvent", "EventStore clean: $isClean")
        }
        Log.i("testTrackSelfDescribingEvent", "Send SelfDescribing event")
        val sdj = SelfDescribingJson("iglu:foo/bar/jsonschema/1-0-0")
        val sdEvent = SelfDescribing(sdj)
        val eventId = Companion.tracker!!.track(sdEvent)
        val req = mockWebServer.takeRequest(60, TimeUnit.SECONDS)
        Assert.assertNotNull(req)
        val reqCount = mockWebServer.requestCount
        Assert.assertEquals(1, reqCount.toLong())
        val payload = JSONObject(req!!.body.readUtf8())
        Assert.assertEquals(2, payload.length().toLong())
        Assert.assertEquals(
            "iglu:com.snowplowanalytics.snowplow/payload_data/jsonschema/1-0-4",
            payload.getString("schema")
        )
        val data = payload.getJSONArray("data")
        Assert.assertEquals(1, data.length().toLong())
        val event = data.getJSONObject(0)
        Assert.assertEquals("ue", event.getString(Parameters.EVENT))
        Assert.assertFalse(event.has(Parameters.UNSTRUCTURED_ENCODED))
        Assert.assertEquals(eventId.toString(), event.getString(Parameters.EID))
        mockWebServer.shutdown()
    }

    @Test
    @Throws(Exception::class)
    fun testTrackWithNoContext() {
        threadCount = 30
        shutdown()
        val namespace = "myNamespace"
        TestUtils.createSessionSharedPreferences(context, namespace)
        val mockWebServer = getMockServer(1)
        var emitter: Emitter? = null
        val builder =
            Consumer { emitterArg: Emitter -> emitterArg.bufferOption = BufferOption.Single }
        try {
            emitter = Emitter(context, getMockServerURI(mockWebServer)!!, builder)
        } catch (e: Exception) {
            e.printStackTrace()
            Assert.fail("Exception on Emitter creation")
        }
        val trackerBuilder = Consumer { tracker: Tracker ->
            tracker.base64Encoded = false
            tracker.logLevel = LogLevel.VERBOSE
            tracker.sessionContext = false
            tracker.platformContextEnabled = false
            tracker.screenContext = false
            tracker.geoLocationContext = false
            tracker.installAutotracking = false
            tracker.screenViewAutotracking = false
        }
        Companion.tracker =
            Tracker(emitter!!, namespace, "testTrackWithNoContext", context, trackerBuilder)
        Log.i("testTrackWithNoContext", "Send ScreenView event")
        Companion.tracker!!.track(ScreenView("name"))
        val req = mockWebServer.takeRequest(60, TimeUnit.SECONDS)
        Assert.assertNotNull(req)
        val reqCount = mockWebServer.requestCount
        Assert.assertEquals(1, reqCount.toLong())
        val payload = JSONObject(req!!.body.readUtf8())
        Assert.assertEquals(2, payload.length().toLong())
        Assert.assertEquals(
            "iglu:com.snowplowanalytics.snowplow/payload_data/jsonschema/1-0-4",
            payload.getString("schema")
        )
        val data = payload.getJSONArray("data")
        Assert.assertEquals(1, data.length().toLong())
        val event = data.getJSONObject(0)
        var found = true
        try {
            val co = event.getString(Parameters.CONTEXT)
            Log.e("testTrackWithNoContext", "Unexpected event: $event")
            Assert.fail(co)
        } catch (e: Exception) {
            found = false
        }
        Assert.assertFalse(found)
        mockWebServer.shutdown()
    }

    @Test
    @Throws(Exception::class)
    fun testTrackWithoutDataCollection() {
        threadCount = 30
        shutdown()
        val namespace = "myNamespace"
        TestUtils.createSessionSharedPreferences(context, namespace)
        val mockWebServer = getMockServer(1)
        val builder = Consumer { emitter: Emitter -> emitter.bufferOption = BufferOption.Single }
        val emitter = Emitter(
            context, getMockServerURI(mockWebServer)!!, builder
        )
        val trackerBuilder = Consumer { tracker: Tracker ->
            tracker.base64Encoded = false
            tracker.logLevel = LogLevel.VERBOSE
            tracker.sessionContext = false
            tracker.platformContextEnabled = false
            tracker.geoLocationContext = false
            tracker.installAutotracking = false
            tracker.exceptionAutotracking = false
            tracker.screenViewAutotracking = false
        }
        Companion.tracker = Tracker(emitter, namespace, "myAppId", context, trackerBuilder)
        Companion.tracker!!.pauseEventTracking()
        val eventId = Companion.tracker!!.track(ScreenView("name"))
        Assert.assertNull(eventId)
        val req = mockWebServer.takeRequest(2, TimeUnit.SECONDS)
        Assert.assertEquals(0, Companion.tracker!!.emitter.eventStore!!.size())
        Assert.assertNull(req)
        mockWebServer.shutdown()
    }

    @Test
    @Throws(Exception::class)
    fun testTrackWithSession() {
        threadCount = 30
        shutdown()
        val namespace = "myNamespace"
        TestUtils.createSessionSharedPreferences(context, namespace)
        val mockWebServer = getMockServer(1)
        val builder = Consumer { emitter: Emitter -> emitter.bufferOption = BufferOption.Single }
        val emitter = Emitter(
            context, getMockServerURI(mockWebServer)!!, builder
        )
        val trackerBuilder = Consumer { tracker: Tracker ->
            tracker.base64Encoded = false
            tracker.logLevel = LogLevel.VERBOSE
            tracker.sessionContext = true
            tracker.platformContextEnabled = false
            tracker.geoLocationContext = false
            tracker.installAutotracking = false
            tracker.exceptionAutotracking = false
            tracker.screenViewAutotracking = false
            tracker.foregroundTimeout = 5
            tracker.backgroundTimeout = 5
        }
        Companion.tracker = Tracker(emitter, namespace, "myAppId", context, trackerBuilder)
        Assert.assertNotNull(Companion.tracker!!.session)
        Companion.tracker!!.resumeSessionChecking()
        Thread.sleep(2000)
        Companion.tracker!!.pauseSessionChecking()
        mockWebServer.shutdown()
    }

    @Test
    fun testTrackScreenView() {
        val namespace = "myNamespace"
        TestUtils.createSessionSharedPreferences(context, namespace)
        val builder = Consumer { emitter: Emitter -> emitter.bufferOption = BufferOption.Single }
        val emitter = Emitter(
            context, "fake-uri", builder
        )
        val trackerBuilder = Consumer { tracker: Tracker ->
            tracker.base64Encoded = false
            tracker.logLevel = LogLevel.VERBOSE
            tracker.screenContext = true
            tracker.sessionContext = false
            tracker.platformContextEnabled = false
            tracker.geoLocationContext = false
            tracker.installAutotracking = false
            tracker.exceptionAutotracking = false
            tracker.screenViewAutotracking = false
            tracker.foregroundTimeout = 5
            tracker.backgroundTimeout = 5
        }
        Companion.tracker = Tracker(emitter, namespace, "myAppId", context, trackerBuilder)
        val screenState = Companion.tracker!!.getScreenState()
        Assert.assertNotNull(screenState)
        var screenStateMapWrapper: Map<String, Any?> = screenState!!.getCurrentScreen(true).map
        var screenStateMap = screenStateMapWrapper[Parameters.DATA] as Map<String?, Any?>?
        Assert.assertEquals("Unknown", screenStateMap!![Parameters.SCREEN_NAME])

        // Send screenView
        var screenView = ScreenView("screen1")
        val screenId = screenView.dataPayload["id"] as String?
        val eventId1 = Companion.tracker!!.track(screenView)
        screenStateMapWrapper = Companion.tracker!!.getScreenState()!!.getCurrentScreen(true).map
        screenStateMap = screenStateMapWrapper[Parameters.DATA] as Map<String?, Any?>?
        Assert.assertEquals("screen1", screenStateMap!![Parameters.SCREEN_NAME])
        Assert.assertEquals(screenId, screenStateMap[Parameters.SCREEN_ID])

        // Send another screenView
        screenView = ScreenView("screen2")
        val screenId1 = screenView.dataPayload["id"] as String?
        val eventId2 = Companion.tracker!!.track(screenView)
        Assert.assertNotEquals(eventId1.toString(), eventId2.toString())
    }

    @Test
    fun testTrackUncaughtException() {
        val namespace = "myNamespace"
        TestUtils.createSessionSharedPreferences(context, namespace)
        Thread.setDefaultUncaughtExceptionHandler(
            TestExceptionHandler("Illegal State Exception has been thrown!")
        )
        Assert.assertEquals(
            TestExceptionHandler::class.java,
            Thread.getDefaultUncaughtExceptionHandler().javaClass
        )
        val emitter = Emitter(
            context, "com.acme", null
        )
        val trackerBuilder = Consumer { tracker: Tracker ->
            tracker.base64Encoded = false
            tracker.logLevel = LogLevel.VERBOSE
            tracker.installAutotracking = false
            tracker.exceptionAutotracking = true
            tracker.screenViewAutotracking = false
        }
        Companion.tracker = Tracker(emitter, namespace, "myAppId", context, trackerBuilder)
        Assert.assertTrue(Companion.tracker!!.exceptionAutotracking)
        Assert.assertEquals(
            ExceptionHandler::class.java,
            Thread.getDefaultUncaughtExceptionHandler().javaClass
        )
    }

    @Test
    fun testExceptionHandler() {
        val namespace = "myNamespace"
        TestUtils.createSessionSharedPreferences(context, namespace)
        val handler = TestExceptionHandler("Illegal State Exception has been thrown!")
        Thread.setDefaultUncaughtExceptionHandler(handler)
        Assert.assertEquals(
            TestExceptionHandler::class.java,
            Thread.getDefaultUncaughtExceptionHandler().javaClass
        )
        val emitter = Emitter(
            context, "com.acme", null
        )
        val trackerBuilder = Consumer { tracker: Tracker ->
            tracker.base64Encoded = false
            tracker.logLevel = LogLevel.VERBOSE
            tracker.installAutotracking = false
            tracker.exceptionAutotracking = false
            tracker.screenViewAutotracking = false
        }
        Companion.tracker = Tracker(emitter, namespace, "myAppId", context, trackerBuilder)
        val handler1 = ExceptionHandler()
        Thread.setDefaultUncaughtExceptionHandler(handler1)
        Assert.assertEquals(
            ExceptionHandler::class.java,
            Thread.getDefaultUncaughtExceptionHandler().javaClass
        )
        handler1.uncaughtException(
            Thread.currentThread(),
            Throwable("Illegal State Exception has been thrown!")
        )
    }

    @Test
    fun testStartsNewSessionWhenChangingAnonymousTracking() {
        val builder = Consumer { emitter: Emitter -> emitter.bufferOption = BufferOption.Single }
        val emitter = Emitter(
            context, "fake-uri", builder
        )
        emitter.pauseEmit()
        val trackerBuilder = Consumer { tracker: Tracker ->
            tracker.base64Encoded = false
            tracker.logLevel = LogLevel.VERBOSE
            tracker.sessionContext = true
            tracker.installAutotracking = false
            tracker.exceptionAutotracking = false
            tracker.screenViewAutotracking = false
            tracker.foregroundTimeout = 5
            tracker.backgroundTimeout = 5
        }
        Companion.tracker = Tracker(emitter, "ns", "myAppId", context, trackerBuilder)
        Companion.tracker!!.track(Structured("c", "a"))
        val sessionIdStart = Companion.tracker!!.session!!.state!!.sessionId
        Companion.tracker!!.userAnonymisation = true
        Companion.tracker!!.track(Structured("c", "a"))
        val sessionIdAnonymous = Companion.tracker!!.session!!.state!!.sessionId
        Assert.assertNotEquals(sessionIdStart, sessionIdAnonymous)
        Companion.tracker!!.userAnonymisation = false
        Companion.tracker!!.track(Structured("c", "a"))
        val sessionIdNotAnonymous = Companion.tracker!!.session!!.state!!.sessionId
        Assert.assertNotEquals(sessionIdAnonymous, sessionIdNotAnonymous)
    }

    class TestExceptionHandler(private val expectedMessage: String) :
        Thread.UncaughtExceptionHandler {
        override fun uncaughtException(t: Thread, e: Throwable) {
            Assert.assertEquals(expectedMessage, e.message)
        }
    }

    // Mock Server
    @Throws(IOException::class)
    fun getMockServer(count: Int): MockWebServer {
        val mockServer = MockWebServer()
        mockServer.start()
        val mockResponse = MockResponse().setResponseCode(200)
        for (i in 0 until count) {
            mockServer.enqueue(mockResponse)
        }
        return mockServer
    }

    @SuppressLint("DefaultLocale")
    fun getMockServerURI(mockServer: MockWebServer?): String? {
        return if (mockServer != null) {
            String.format("%s:%d", mockServer.hostName, mockServer.port)
        } else null
    }

    companion object {
        private var tracker: Tracker? = null
    }
}
